packages/utils/src/split.ts

```diff
@@ -0,0 +1,57 @@
+import { Observable, GroupedObservable, Subject, ReplaySubject } from "rxjs"
+import { shareReplay } from "rxjs/operators"
+
+export function split<K, T>(
+  keySelector: (value: T) => K,
+): (stream: Observable<T>) => Observable<GroupedObservable<K, T>>
+
+export function split<K, T, R>(
+  keySelector: (value: T) => K,
+  streamSelector: (grouped: Observable<T>, key: K) => Observable<R>,
+): (stream: Observable<T>) => Observable<GroupedObservable<K, R>>
+
+export function split<K, T, R>(
+  keySelector: (value: T) => K,
+  streamSelector?: (grouped: Observable<T>, key: K) => Observable<R>,
+) {
+  return (stream: Observable<T>) =>
+    new Observable<GroupedObservable<K, R>>((subscriber) => {
+      const groups: Map<K, Subject<T>> = new Map()
+
+      return stream.subscribe(
+        (x) => {
+          const key = keySelector(x)
+          if (groups.has(key)) {
+            return groups.get(key)!.next(x)
+          }
+
+          const subject = streamSelector
+            ? new Subject<T>()
+            : new ReplaySubject<T>(1)
+          groups.set(key, subject)
+
+          const res = (streamSelector
+            ? streamSelector(subject, key).pipe(shareReplay(1))
+            : subject.asObservable()) as GroupedObservable<K, R>
+
+          res.key = key
+          res.subscribe({
+            complete() {
+              groups.delete(key)
+            },
+          })
```

Comment:
This subscription will need cleaning up in the following scenario:

```js
cold('1-2--3-4').pipe(
  split(v => v % 2),
  take(2)
).subscribe()
```

As there's nothing completing the inner subject.

I was thinking this is only relevant when you don't pass in a `streamSelector` parameter, because if you do then you have control when to stop it. But I can't think of an elegant way to solve this same example using streamSelector.... `takeUntil(what?)`: At that point we don't know the logic of the consumer (when will it unsubscribe?).

Reply:
Good catch! I've addressed that and I've added [a test for this case](https://github.com/re-rxjs/react-rxjs/pull/79/files#diff-0c49ff00c654149cec7eaa7af1e52bdcR24-R37).